### DISCUSSION TITLE
Swarm using old version of Python on on my Mac - ComfyMath returns a TypeAlias ImportError on startup

### DISCUSSION THREAD

**Discussion by mozthefox** [Stability-AI/StableSwarmUI#241] [General]
Hello ☺️

I'm experiencing an issue with Swarm using the wrong version of Python on macOS Sonoma (I have an M1 Max MacBook Pro). It keeps defaulting to Python 3.9 despite my environment being set up for Python 3.10.11 via `pyenv`. Swarm works absolutely fine for me however I am trying to use the [ComfyMath](https://github.com/evanspearman/ComfyMath) nodes, but this leads to the following `ImportError` when starting up Swarm related to `TypeAlias`, which, I believe (?), is not supported in Python 3.9.

`12:08:33.534 [Warning] ComfyUI-0 on port 7821 stdout: ImportError: cannot import name 'TypeAlias' from 'typing' (/Library/Developer/CommandLineTools/Library/Frameworks/Python3.framework/Versions/3.9/lib/python3.9/typing.py)`

I'm not sure why Swarm/ComfyUI is using that version of Python - I believe the folder location suggests it's the command line tools from Xcode(?). Python and .NET aren't my bag, unfortunately, so apologies if my understanding is wrong or I've missed something really obvious. Anyways, here are the steps I've taken up to this point:

- Installed Swarm following the official guide, substituting Step 4 with the official Microsoft .NET 8 Mac Arm64 installer
- I use `pyenv` to manage Python installs and it's currently set to use Python 3.10.11
- I've updated `.zshrc` with the recommended `pyenv` initialization commands, and verified the correct version with `python3 --version`
- I tried modifying the `launch-macos.sh` script to export the `pyenv` Python path before execution, but this did nothing.

Despite these efforts, the server output indicates Swarm, dotnet, or maybe a subsystem, is still invoking Python 3.9.

Does anyone know how I can get ComfyMath working or get Swarm to use the Python version setup in `pyenv`? Any insights or suggestions would be greatly appreciated.

Thank you!
James

**Comment by mcmonkey4eva** [comment]
If you did the switch *after* installing, it will be stuck on the old one - the install environment has its own python version that links back to the global install env that was used to create it.

Unfortunately, to my knowledge, migrating an env between python versions is annoyingly difficult - the easiest thing to do is just nuke & rebuild.

If you have a bit of basic technical experience you can open the comfy folder under `dlbackend/ComfyUI` and delete the `venv` folder, then:
(first double check `python3 --version` still outputs correctly as 3.10)
```sh
python3 -s -m venv venv
. venv/bin/activate
python3 -s -m pip install torch torchvision torchaudio --extra-index-url https://download.pytorch.org/whl/cu121
python3 -s -m pip install -r requirements.txt
```
Then also install any relevant custom node requirements too

If you have a bit more technical experience you can do the `pip freeze` thing to copy exact packages but that'll go wonky if you're not careful

ps - the reason for the third party package of dotnet isn't for net8, it's for net _7_ -- currently Swarm uses 7 and is transitioning to 8 soon, thus installers trying to preinclude 8 in advance to reduce rate of issues once the changeover is done. Mac's packages just hardswapped to 8 without an easy route to install the old 7, and for some reason dotnet8 doesn't just support net7, so a thirdparty brew package was needed.

**Reply by mozthefox** [reply]
You are a genius. Thank you so so much. Yeah, I figured out about the Python issue after installing Swarm. Nuking the venv folder and setting it back up again worked. ComfyMath now loads without issue.

I really appreciate the time you took to explain everything. That all makes perfect sense. Would never have found that without your guidance.

Thanks again,
James
